### ISSUE TITLE
addForeignKey() on lookup tables leads to problems

### ISSUE THREAD

**Issue by aecker** [datajoint/datajoint-matlab#42]
# Problem

I noticed that adding foreign keys via `addForeignKey()` on a table of type `lookup` leads to various problems. In the best case it results in a MySQL error and fails to add the foreign key; in the worst case it creates the foreign key, but any future changes will fail and the only solution is to drop the table and re-create it.
# Reason

When MySQL automatically creates the name for the foreign key (usually `<tablename>_ibfk_x`), the `#` character that is prepended to the table name for lookup tables creates problems. The following seems to happen:
1. When the foreign key constraint is declared during table declaration everything works fine.
2. When adding a foreign key via `ALTER TABLE` (as is done in `addForeignKey()`) the `#` is translated to `@0023`. If table `foo` has already a foreign key named `#foo_ibfk_1`, the new foreign key gets named `@0023foo_ibfk_1` (note the `1` instead of `2` as should be expected). 
3. Now, depending on the MySQL server version, this either (a) creates an error 1050 because the two are interpreted as being named identically or (b) works fine, creates the foreign key, but prevents any future changes to the table, which will fail because apparently two identically named keys exist.

Bottom line: there seem to be situations in which MySQL thinks `#` is the same as `@0023` and other situations where it doesn't. This is either a bug in MySQL or `#` is not supposed to be part of a table name (or both).
# Suggested solution

Ideally we probably shouldn't use # as a prefix, but that will break backwards compatibility. As a workaround, we could manually name the foreign key in `addForeignKey()` as follows:

﻿﻿`ALTER TABLE 'foo'.'#bar' ADD CONSTRAINT '#bar_ibfk_2' FOREIGN KEY ...`

I have tested this approach and it works fine. It would require some logistics reading out the existing foreign keys to determine the number or, alternatively, we could just create a random name for each new key.

**Comment by dimitri-yatsenko** [comment]
we could start using a new prefix in new tables while keeping backward compatibility.  What prefix would you suggest?  

We could also explicitly name the foreign keys.

**Comment by aecker** [comment]
We could just drop all the prefixes. What's their purpose? The table names are neither exposed to the user nor can there be two identically named tables of different tiers, right?

We do have to name the foreign keys explicitly if we want to fix this (pretty nasty) bug for existing tables. The question is whether we care about the names or just go with a randomly generated one (I would vote for a random one due to simplicity of implementation).

**Comment by dimitri-yatsenko** [comment]
The prefixes are helpful to administrators for backup, etc.  I have my server set up so that only manual tables are backed up daily and imported tables are backed up weekly and computed tables are never backed up.

**Comment by dimitri-yatsenko** [comment]
I support dropping the prefixes.  It was Florian who asked for them around 2011 and I have not heard about them since. Any objections?

**Comment by dimitri-yatsenko** [comment]
what's the verdict on this?
